plugins/catalog-backend-module-puppetdb/src/providers/PuppetDbEntityProvider.ts

```diff
@@ -0,0 +1,236 @@
+  Entity,
+} from '@backstage/catalog-model/';
+import { merge } from 'lodash';
+import { readPuppetNodes } from '../puppet/read';
+import { ENDPOINT_NODES } from '../puppet/constants';
+
+/**
+ * Reads nodes from [PuppetDB](https://www.puppet.com/docs/puppet/6/puppetdb_overview.html)
+ * based on the provided query and registers them as Resource entities in the catalog.
+ *
+ * @public
+ */
+export class PuppetDbEntityProvider implements EntityProvider {
+  private readonly config: PuppetDbEntityProviderConfig;
+  private readonly logger: Logger;
+  private readonly scheduleFn: () => Promise<void>;
+  private readonly transformer: ResourceTransformer;
+  private connection?: EntityProviderConnection;
+
+  /**
+   * Creates instances of {@link PuppetDbEntityProvider} from a configuration.
+   *
+   * @param config - The configuration to read provider information from.
+   * @param deps - The dependencies for  {@link PuppetDbEntityProvider}.
+   *
+   * @returns A list of {@link PuppetDbEntityProvider} instances.
+   */
+  static fromConfig(
+    config: Config,
+    deps: {
+      logger: Logger;
+      schedule?: TaskRunner;
+      scheduler?: PluginTaskScheduler;
+      transformer?: ResourceTransformer;
+    },
+  ): PuppetDbEntityProvider[] {
+    if (!deps.schedule && !deps.scheduler) {
+      throw new Error('Either schedule or scheduler must be provided.');
+    }
+
+    return readProviderConfigs(config).map(providerConfig => {
+      if (!deps.schedule && !providerConfig.schedule) {
+        throw new Error(
+          `No schedule provided neither via code nor config for puppet-provider:${providerConfig.id}.`,
+        );
+      }
+
+      const taskRunner =
+        deps.schedule ??
+        deps.scheduler!.createScheduledTaskRunner(providerConfig.schedule!);
+
+      const transformer = deps.transformer ?? defaultResourceTransformer;
+
+      return new PuppetDbEntityProvider(
+        providerConfig,
+        deps.logger,
+        taskRunner,
+        transformer,
+      );
+    });
+  }
+
+  /**
+   * Creates an instance of {@link PuppetDbEntityProvider}.
+   *
+   * @param config - Configuration of the provider.
+   * @param logger - The instance of a {@link Logger}.
+   * @param taskRunner - The instance of {@link TaskRunner}.
+   * @param transformer - A {@link ResourceTransformer} function.
+   *
+   * @private
+   */
+  private constructor(
+    config: PuppetDbEntityProviderConfig,
+    logger: Logger,
+    taskRunner: TaskRunner,
+    transformer: ResourceTransformer,
+  ) {
+    this.config = config;
+    this.logger = logger.child({
+      target: this.getProviderName(),
+    });
+    this.scheduleFn = this.createScheduleFn(taskRunner);
+    this.transformer = transformer;
+  }
+
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.connect} */
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.connection = connection;
+    await this.scheduleFn();
```

Comment:
You don't need to setup the task after connection is created. The scheduler won't run any tasks before entry providers are setup. You can create the task in fromConfig. The structure of your plug-in becomes non-conditional, but this will work too.

Reply:
Thanks!

As for the scheduler task, I did this for consistency, as it seems [all other catalog backend plugins are doing the same](https://github.com/search?q=repo%3Abackstage%2Fbackstage%20await%20this.scheduleFn()%3B&type=code).

Should I move `await this.scheduleFn();` to `fromConfig` instead?

Reply:
No, it's fine. That's the correct way